finat/direct_serendipity.py

```diff
@@ -0,0 +1,371 @@
+
+    nodes1d = [sympy.Rational(i, r) for i in range(1, r)]
+
+    s = sympy.Symbol('s')
+
+    # for each edge:
+    # I need its adjacent two edges
+    # and its opposite edge
+    # and its "tunnel R" RH or RV
+    opposite_edges = {0: 1, 1: 0, 2: 3, 3: 2}
+    adjacent_edges = {0: (2, 3), 1: (2, 3), 2: (0, 1), 3: (0, 1)}
+    tunnel_R_edges = {0: RH, 1: RH, 2: RV, 3: RV}
+
+    edge_nodes = []
+    for ed in range(4):
+        ((v0x, v0y), (v1x, v1y)) = vs[ct[1][ed], :]
+        delx = v1x - v0x
+        dely = v1y - v0y
+        edge_nodes.append([(v0x+nd*delx, v0y+nd*dely) for nd in nodes1d])
+
+    # subtracts off the value of function at internal nodes times those
+    # internal basis functions
+    def nodalize(f):
+        foo = f
+        for (bf, nd) in zip(internal_bfs, internal_nodes):
+            foo = foo - f.subs(xx, nd) * bf
+        return foo
+
+    edge_bfs = []
+    if r == 2:
+        for ed in range(4):
+            lamadj0 = lams[adjacent_edges[ed][0]]
+            lamadj1 = lams[adjacent_edges[ed][1]]
+            ephi = lamadj0 * lamadj1 * Rs[ed]
+            phi = nodalize(ephi) / ephi.subs(xysub(xx, xstars[ed]))
+            edge_bfs.append([phi])
+    else:
+        for ed in range(4):
+            ((v0x, v0y), (v1x, v1y)) = vs[ct[1][ed], :]
+            Rcur = tunnel_R_edges[ed]
+            lam_op = lams[opposite_edges[ed]]
+
+            edge_bfs_cur = []
+
+            for i in range(len(nodes1d)):
+                # strike out i:th node
+                idcs = [j for j in range(len(nodes1d)) if i != j]
+                nodes1d_cur = [nodes1d[j] for j in idcs]
+                edge_nodes_cur = [edge_nodes[ed][j]
+                                  for j in idcs]
+
+                # construct the 1d interpolation with remaining nodes
+                pvals = []
+                for nd in edge_nodes_cur:
+                    sub = xysub(xx, nd)
+                    pval_cur = (-1 * Rcur.subs(sub)**(r-2)
+                                / lam_op.subs(sub))
+                    pvals.append(pval_cur)
+
+                ptilde = newton_poly(nodes1d_cur, pvals, s)
+                xt = xx @ ts[ed]
+                vt0 = numpy.asarray((v0x, v0y)) @ ts[ed]
+                vt1 = numpy.asarray((v1x, v1y)) @ ts[ed]
+                p = ptilde.subs({s: (xt-vt0) / (vt1-vt0)})
+
+                prebf = (lams[adjacent_edges[ed][0]]
+                         * lams[adjacent_edges[ed][1]]
+                         * (lams[opposite_edges[ed]] * p
+                            + Rcur**(r-2) * Rs[ed]))
+
+                bfcur = (nodalize(prebf)
+                         / prebf.subs(xysub(xx, edge_nodes[ed][i])))
+                edge_bfs_cur.append(bfcur)
+
+            edge_bfs.append(edge_bfs_cur)
+
+    # vertex basis functions
+    vertex_to_adj_edges = {0: (0, 2), 1: (0, 3), 2: (1, 2), 3: (1, 3)}
+    vertex_to_off_edges = {0: (1, 3), 1: (1, 2), 2: (0, 3), 3: (0, 2)}
```

Comment:
This encodes information about the fiat cell topology I think, can you deduce it from that, rather than hardcoding here (just in case we ever change anything).